File: MiniDOMOffice/ADVs/ADVConfig.vb
Imports minidom
Imports minidom.Databases
Imports minidom.Sistema
Imports minidom.Anagrafica



Partial Public Class ADV

    Public NotInheritable Class ADVConfig
        Inherits DBObject

        Private m_BannedEMailAddresses As New ADVBannedAddressCollection
        Private m_BannedFaxAddresses As New ADVBannedAddressCollection
        Private m_BannedSMSAddresses As New ADVBannedAddressCollection
        Private m_BannedWebAddresses As New ADVBannedAddressCollection

        Public Sub New()
        End Sub

        Public ReadOnly Property BannedEMailAddresses As ADVBannedAddressCollection
            Get
                Return Me.m_BannedEMailAddresses
            End Get
        End Property

        Public ReadOnly Property BannedFaxAddresses As ADVBannedAddressCollection
            Get
                Return Me.m_BannedFaxAddresses
            End Get
        End Property

        Public ReadOnly Property BannedSMSAddresses As ADVBannedAddressCollection
            Get
                Return Me.m_BannedSMSAddresses
            End Get
        End Property

        Public ReadOnly Property BannedWebAddresses As ADVBannedAddressCollection
            Get
                Return Me.m_BannedWebAddresses
            End Get
        End Property

        Protected Overrides Function SaveToDatabase(dbConn As CDBConnection, force As Boolean) As Boolean
            Dim ret As Boolean = MyBase.SaveToDatabase(dbConn, force)
            SetCOnfig(Me)
            Return ret
        End Function

        Public Overrides Function GetModule() As CModule
            Return Nothing
        End Function

        Friend Sub Load()
            Dim dbSQL As String = "SELECT * FROM [tbl_ADVConfig] ORDER BY [ID] ASC"
            Dim dbRis As System.Data.IDataReader = ADV.Database.ExecuteReader(dbSQL)
            If (dbRis.Read) Then
                ADV.Database.Load(Me, dbRis)
            End If
            dbRis.Dispose()
        End Sub

        Protected Overrides Function SaveToRecordset(writer As DBWriter) As Boolean
            Me.m_BannedEMailAddresses.Write(writer, "BannedEMailAddresses")
            Me.m_BannedFaxAddresses.Write(writer, "BannedFaxAddresses")
            Me.m_BannedSMSAddresses.Write(writer, "BannedSMSAddresses")
            Me.m_BannedWebAddresses.Write(writer, "BannedWebAddresses")
            Return MyBase.SaveToRecordset(writer)
        End Function

        Protected Overrides Function LoadFromRecordset(reader As DBReader) As Boolean
            Me.m_BannedEMailAddresses.Read(reader, "BannedEMailAddresses")
            Me.m_BannedFaxAddresses.Read(reader, "BannedFaxAddresses")
            Me.m_BannedSMSAddresses.Read(reader, "BannedSMSAddresses")
            Me.m_BannedWebAddresses.Read(reader, "BannedWebAddresses")
            Return MyBase.LoadFromRecordset(reader)
        End Function

        Public Overrides Function GetTableName() As String
            Return "tbl_ADVConfig"
        End Function

        Protected Overrides Function GetConnection() As CDBConnection
            Return ADV.Database
        End Function

        Protected Overrides Sub XMLSerialize(writer As XML.XMLWriter)
            MyBase.XMLSerialize(writer)
            writer.WriteTag("BannedEMailAddresses", Me.m_BannedEMailAddresses)
            writer.WriteTag("BannedFaxAddresses", Me.m_BannedFaxAddresses)
            writer.WriteTag("BannedSMSAddresses", Me.m_BannedSMSAddresses)
            writer.WriteTag("BannedWebAddresses", Me.m_BannedWebAddresses)
        End Sub

        Protected Overrides Sub SetFieldInternal(fieldName As String, fieldValue As Object)
            Select Case fieldName
                Case "BannedEMailAddresses" : Me.m_BannedEMailAddresses = fieldValue
                Case "BannedFaxAddresses" : Me.m_BannedFaxAddresses = fieldValue
                Case "BannedSMSAddresses" : Me.m_BannedSMSAddresses = fieldValue
                Case "BannedWebAddresses" : Me.m_BannedWebAddresses = fieldValue
                Case Else : MyBase.SetFieldInternal(fieldName, fieldValue)
            End Select
        End Sub

    End Class

    ''' <summary>
    ''' Evento generato quando viene modificata la configurazione
    ''' </summary>
    ''' <param name="sender"></param>
    ''' <param name="e"></param>
    ''' <remarks></remarks>
    Public Shared Event ConfigurationChanged(ByVal sender As Object, ByVal e As System.EventArgs)


    Private Shared m_Config As ADVConfig

    ''' <summary>
    ''' Oggetto che contiene alcuni parametri relativi alla configurazione del modulo
    ''' </summary>
    ''' <value></value>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public Shared ReadOnly Property Configuration As ADVConfig
        Get
            If m_Config Is Nothing Then
                m_Config = New ADVConfig
                m_Config.Load()
            End If
            Return m_Config
        End Get
    End Property

    Friend Shared Sub SetConfig(ByVal value As ADVConfig)
        m_Config = value
        RaiseEvent ConfigurationChanged(Nothing, New System.EventArgs)
    End Sub


End Class

File: MiniDOMOffice/ADVs/Config/ADVConfig.cs
using System;
using DMD;
using DMD.XML;
using minidom;
using static minidom.Sistema;

namespace minidom
{
    public partial class ADV
    {

        /// <summary>
        /// Configurazione delle campagne pubblicitarie
        /// </summary>
        [Serializable]
        public sealed class ADVConfig 
            : DMD.XML.DMDBaseXMLObject 
        {
            private ADVBannedAddressCollection m_BannedEMailAddresses = new ADVBannedAddressCollection();
            private ADVBannedAddressCollection m_BannedFaxAddresses = new ADVBannedAddressCollection();
            private ADVBannedAddressCollection m_BannedSMSAddresses = new ADVBannedAddressCollection();
            private ADVBannedAddressCollection m_BannedWebAddresses = new ADVBannedAddressCollection();

            /// <summary>
            /// Costruttore
            /// </summary>
            public ADVConfig()
            {
            }

            /// <summary>
            /// Indirizzi email bannati
            /// </summary>
            public ADVBannedAddressCollection BannedEMailAddresses
            {
                get
                {
                    return m_BannedEMailAddresses;
                }
            }

            /// <summary>
            /// Indirizzi fax bannati
            /// </summary>
            public ADVBannedAddressCollection BannedFaxAddresses
            {
                get
                {
                    return m_BannedFaxAddresses;
                }
            }

            /// <summary>
            /// Indirizzi sms bannati
            /// </summary>
            public ADVBannedAddressCollection BannedSMSAddresses
            {
                get
                {
                    return m_BannedSMSAddresses;
                }
            }

            /// <summary>
            /// Indirizzi web bannati
            /// </summary>
            public ADVBannedAddressCollection BannedWebAddresses
            {
                get
                {
                    return m_BannedWebAddresses;
                }
            }

            //protected override bool SaveToDatabase(Databases.CDBConnection dbConn, bool force)
            //{
            //    bool ret = base.SaveToDatabase(dbConn, force);
            //    SetConfig(this);
            //    return ret;
            //}

            //public override CModulesClass GetModule()
            //{
            //    return null;
            //}

            //internal void Load()
            //{
            //    string dbSQL = "SELECT * FROM [tbl_ADVConfig] ORDER BY [ID] ASC";
            //    var dbRis = Database.ExecuteReader(dbSQL);
            //    if (dbRis.Read())
            //    {
            //        Database.Load(this, dbRis);
            //    }

            //    dbRis.Dispose();
            //}

            //protected override bool SaveToRecordset(DBWriter writer)
            //{
            //    m_BannedEMailAddresses.Write(writer, "BannedEMailAddresses");
            //    m_BannedFaxAddresses.Write(writer, "BannedFaxAddresses");
            //    m_BannedSMSAddresses.Write(writer, "BannedSMSAddresses");
            //    m_BannedWebAddresses.Write(writer, "BannedWebAddresses");
            //    return base.SaveToRecordset(writer);
            //}

            //protected override bool LoadFromRecordset(DBReader reader)
            //{
            //    m_BannedEMailAddresses.Read(reader, "BannedEMailAddresses");
            //    m_BannedFaxAddresses.Read(reader, "BannedFaxAddresses");
            //    m_BannedSMSAddresses.Read(reader, "BannedSMSAddresses");
            //    m_BannedWebAddresses.Read(reader, "BannedWebAddresses");
            //    return base.LoadFromRecordset(reader);
            //}

            //public override string GetTableName()
            //{
            //    return "tbl_ADVConfig";
            //}

            //protected override Databases.CDBConnection GetConnection()
            //{
            //    return Database;
            //}

            /// <summary>
            /// Serializzazione xml
            /// </summary>
            /// <param name="writer"></param>
            protected override void XMLSerialize(XMLWriter writer)
            {
                base.XMLSerialize(writer);
                writer.WriteTag("BannedEMailAddresses", m_BannedEMailAddresses);
                writer.WriteTag("BannedFaxAddresses", m_BannedFaxAddresses);
                writer.WriteTag("BannedSMSAddresses", m_BannedSMSAddresses);
                writer.WriteTag("BannedWebAddresses", m_BannedWebAddresses);
            }

            /// <summary>
            /// Deserializzazione xml
            /// </summary>
            /// <param name="fieldName"></param>
            /// <param name="fieldValue"></param>
            protected override void SetFieldInternal(string fieldName, object fieldValue)
            {
                switch (fieldName ?? "")
                {
                    case "BannedEMailAddresses":
                        {
                            m_BannedEMailAddresses = (ADVBannedAddressCollection)fieldValue;
                            break;
                        }

                    case "BannedFaxAddresses":
                        {
                            m_BannedFaxAddresses = (ADVBannedAddressCollection)fieldValue;
                            break;
                        }

                    case "BannedSMSAddresses":
                        {
                            m_BannedSMSAddresses = (ADVBannedAddressCollection)fieldValue;
                            break;
                        }

                    case "BannedWebAddresses":
                        {
                            m_BannedWebAddresses = (ADVBannedAddressCollection)fieldValue;
                            break;
                        }

                    default:
                        {
                            base.SetFieldInternal(fieldName, fieldValue);
                            break;
                        }
                }
            }
        }

        /// <summary>
        /// Evento generato quando viene modificata la configurazione
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        /// <remarks></remarks>
        public static event ConfigurationChangedEventHandler ConfigurationChanged;

        public delegate void ConfigurationChangedEventHandler(object sender, EventArgs e);

        private static ADVConfig m_Config;

        /// <summary>
        /// Oggetto che contiene alcuni parametri relativi alla configurazione del modulo
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
        public static ADVConfig Configuration
        {
            get
            {
                if (m_Config is null)
                {
                    var str = minidom.Sistema.ApplicationContext.Settings.GetValueString("ADV.Config", "");
                    m_Config = DMD.XML.Utils.Deserialize<ADVConfig>(str);
                }

                return m_Config;
            }
        }

        public static void SetConfig(ADVConfig value)
        {
            var str = DMD.XML.Utils.Serialize(value);
            minidom.Sistema.ApplicationContext.Settings.SetValueString("ADV.Config", str);
            m_Config = value;
            ConfigurationChanged?.Invoke(null, new EventArgs());
        }
    }
}